frontend/src/editor/speaker_dropdown.tsx

```diff
@@ -159,94 +145,87 @@ function setSpeaker(editor: AutomergeEditor, path: number[], speaker: string | n
-
-            // No-op to trigger a re-render of the editor so the speaker name gets applied
-            editor.apply({
-              type: 'set_node',
-              path: [0],
-              properties: {},
-              newProperties: {},
-            });
+          onSpeakerSelected={(speakerId) => setSpeaker(editor, node, speakerId)}
+          onNewSpeaker={(speakerName) => {
+            const speakerId = addNewSpeaker(editor, speakerName);
+            setSpeaker(editor, node, speakerId);
           }}
         />,
       );
-    }
-  };
-  const unsetSpeaker = () => setSpeaker(editor, elementPath, null);
+    }, [speaker, node, editor]);
+
+    const renameSpeaker = useCallback(() => {
+      if (speaker !== null && speaker !== undefined) {
+        showModal(
+          <SpeakerNameModal
+            onClose={() => showModal(null)}
+            initialValue={speakerName}
+            selectedCallback={(speakerName: string) => {
+              changeSpeakerName(editor, speaker, speakerName);
+            }}
+          />,
+        );
+      }
+    }, [speaker, speakerName, editor]);
+
+    const unsetSpeaker = useCallback(() => setSpeaker(editor, node, null), [editor, node]);
+
+    return (
+      <Dropdown label={speakerName} className="pr-4">
+        <DropdownSection>
+          <DropdownItem icon={IoIosList} onClick={changeSpeaker}>
+            Change Speaker
+          </DropdownItem>
+          <DropdownItem
+            icon={IoIosCreate}
+            onClick={renameSpeaker}
+            disabled={speaker === null || speaker === undefined}
+          >
+            Rename Speaker
+          </DropdownItem>
+        </DropdownSection>
+        <DropdownSection>
+          <DropdownItem
+            icon={IoIosTrash}
+            onClick={unsetSpeaker}
+            disabled={speaker === null || speaker === undefined}
+          >
+            Unset Speaker
+          </DropdownItem>
+        </DropdownSection>
+      </Dropdown>
+    );
+  },
+);
 
-  return (
-    <Dropdown label={getSpeakerName(paragraph, editor.doc.speaker_names)} className="pr-4">
-      <DropdownSection>
-        <DropdownItem icon={IoIosList} onClick={changeSpeaker}>
-          Change Speaker
-        </DropdownItem>
-        <DropdownItem
-          icon={IoIosCreate}
-          onClick={renameSpeaker}
-          disabled={paragraph.speaker === null || paragraph.speaker === undefined}
-        >
-          Rename Speaker
-        </DropdownItem>
-      </DropdownSection>
-      <DropdownSection>
-        <DropdownItem
-          icon={IoIosTrash}
-          onClick={unsetSpeaker}
-          disabled={paragraph.speaker === null || paragraph.speaker === undefined}
-        >
-          Unset Speaker
-        </DropdownItem>
-      </DropdownSection>
-    </Dropdown>
-  );
+SpeakerDropdownInner.displayName = 'SpeakerDropdownInner';
+
+export function SpeakerDropdown({ paragraph }: { paragraph: Paragraph }) {
+  const editor: Editor = useSlateStatic();
```

Comment:
does this really have the needed rerender guarantees? This component is not rerendered if a speaker name changes?!

Reply:
It does, because editor is only shoved into the `useDocument` (or rather `useSpeakerName{,s}`.
But I'll change those hooks to make that clearer.